### PR TITLE
[Android] Fix issue with minimumDate/maximumDate when using timeZoneOffsetInMinutes

### LINKED ISSUE #518: [Android] timeZoneOffsetInMinutes issue with minimumDate/maximumDate
# Bug report

## Summary

When setting a minimum and maximum UTC date and setting `timeZoneOffsetInMinutes={0}` the Android picker is off by one day, I'm guessing it's using the device timezone somewhere.

## Reproducible sample code

Using the example project, change these lines:
```js
const [minDate, setMaxDate] = useState(new Date(Date.UTC(2021, 9, 2)));
const [maxDate, setMinDate] = useState(new Date(Date.UTC(2021, 9, 4)));
```

And also specify this for the `<DateTimePicker />`:
```jsx
minimumDate={minDate}
maximumDate={maxDate}
```

(iOS Left: No issues, Android Right: Issue)
Screen Shot 2021-10-06 at 9 23 14 AM

## Steps to reproduce

1. Click Show Picker

Describe what you expected to happen:

1. Should only let the user pick 2,3,4 instead, it let me pick 1,2,3 on Android.

## Environment info

`npx react-native info` output:

```bash
System:
    OS: macOS 11.6
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 1.25 GB / 32.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.16.1 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.14.0 - ~/Development/datetimepicker/node_modules/.bin/npm
    Watchman: 2021.06.07.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.10.1 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 15.0, DriverKit 20.4, macOS 11.3, tvOS 15.0, watchOS 8.0
    Android SDK:
      API Levels: 28, 29, 30, 31
      Build Tools: 28.0.3, 29.0.1, 29.0.2, 29.0.3, 30.0.0, 30.0.1, 30.0.2, 30.0.3, 31.0.0
      System Images: android-29 | Google Play Intel x86 Atom, android-30 | Google Play Intel x86 Atom, android-31 | Google Play Intel x86 Atom_64
      Android NDK: Not Found
  IDEs:
    Android Studio: Arctic Fox 2020.3.1 Patch 2 Arctic Fox 2020.3.1 Patch 2
    Xcode: 13.0/13A233 - /usr/bin/xcodebuild
  Languages:
    Java: 12.0.2 - /usr/bin/javac
    Python: 2.7.16 - /usr/bin/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.11.0 => 16.11.0
    react-native: ^0.62.2 => 0.62.3
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

datetimepicker version: 3.5.2

iOS / Android version: iOS 15.0 / Android 10

### PR DESCRIPTION
**Pull request by PaitoAnderson**
Fixes #518 

# Summary

On Android `setMinDate` and `setMaxDate` assumes the datetime passed will be in the device's local timezone, we now account for this by subtracting the minimumDate/maximumDate datetime offsets and also the provided timeZoneOffsetInMinutes.

## Test Plan

I've added a `toggleMinMaxDate` button in the demo project that shows this feature using two UTC min/max dates. Here is the before/after in -4 GMT.

Before:
Screenshot_1633700803
After:
Screenshot_1633700838

### What's required for testing (prerequisites)?

Test using the `toggleMinMaxDate` button in the demo project.

FYI - I was unable to find a good way to pick a date using detox for e2e tests on Android.

### What are the steps to reproduce (after prerequisites)?

Using any negative timezone on the device should show this issue in the demo project.

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ✅     |

## Checklist

- [x] I have tested this on a device and a simulator
- [ ] I added the documentation in `README.md`
- [ ] I updated the typed files (TS and Flow)
- [x] I added a sample use of the API in the example project (`example/App.js`)
- [x] I have added automated tests, either in JS or e2e tests, as applicable

### CODE REVIEW

File: android/src/main/java/com/reactcommunity/rndatetimepicker/RNDatePickerDialogFragment.java [hunk 1]
```diff
@@ -134,7 +143,13 @@ static DatePickerDialog createDialog(
       c.set(Calendar.MINUTE, 0);
       c.set(Calendar.SECOND, 0);
       c.set(Calendar.MILLISECOND, 0);
-      datePicker.setMinDate(c.getTimeInMillis());
+
+      if (timeZoneOffsetInMilliseconds != null) {
+        int offset = TimeZone.getDefault().getOffset(c.getTimeInMillis()) - timeZoneOffsetInMilliseconds;
```
Comment (vonovak) [inline android/src/main/java/com/reactcommunity/rndatetimepicker/RNDatePickerDialogFragment.java:148]: can you please extract this into a function `int getOffset()` that returns 0 or `TimeZone.getDefault().getOffset(c.getTimeInMillis()) - timeZoneOffsetInMilliseconds`?

then reuse the fn here and for datePicker.setMaxDate()

thank you :)

File: android/src/main/java/com/reactcommunity/rndatetimepicker/RNDatePickerDialogFragment.java [hunk 2]
```diff
@@ -125,6 +126,14 @@ static DatePickerDialog createDialog(
 
     final DatePicker datePicker = dialog.getDatePicker();
 
+    Integer timeZoneOffsetInMilliseconds = null;
+    if (args != null && args.containsKey(RNConstants.ARG_TZOFFSET_MINS)) {
+      c.setTimeZone(TimeZone.getTimeZone("GMT"));
+      long timeZoneOffsetInMinutesFallback = args.getLong(RNConstants.ARG_TZOFFSET_MINS);
```
Comment (vonovak) [inline android/src/main/java/com/reactcommunity/rndatetimepicker/RNDatePickerDialogFragment.java:132]: can this logic be xtracted into a separate fn and the `Integer timeZoneOffsetInMilliseconds = ` would be a one-liner?

thank would be nice, thank you

**Review by vonovak — COMMENTED** [review]
hello and thank you for the PR!

I have added some small comments. Also, I'd really appreciate if you could add some e2e tests for this so that we know the android and ios behavior is the same, since the tz offset on android is not working 100%.

Selecting dates on android is kinda pain, the best way I know is in this PR: //github.com/react-native-datetimepicker/datetimepicker/pull/new/fix/use-resource-ids

thank you for your time! :)

**Comment by PaitoAnderson** [comment]
Thanks for reviewing and pointing me in the right direction on the android date picker selection for the e2e test.

File: android/src/main/java/com/reactcommunity/rndatetimepicker/RNDatePickerDialogFragment.java [hunk 3]
```diff
@@ -130,12 +136,30 @@ static DatePickerDialog createDialog(
...
       c.set(Calendar.MILLISECOND, 999);
-      datePicker.setMaxDate(c.getTimeInMillis());
+      datePicker.setMaxDate(c.getTimeInMillis() - getOffset(c, timeZoneOffsetInMilliseconds));
     }
 
     return dialog;
   }
 
+  private static Integer getTimeZoneOffset(Bundle args)
+  {
```
Comment (vonovak) [inline android/src/main/java/com/reactcommunity/rndatetimepicker/RNDatePickerDialogFragment.java:146]:
```suggestion
  private static Integer getTimeZoneOffset(Bundle args) {
```

File: example/App.js [hunk 4]
```diff
@@ -111,6 +114,17 @@ export const App = () => {
...
+    setMinimumDate(
+      maximumDate ? undefined : sourceMoment.utc().startOf('day').toDate(),
+    );
+    setMaximumDate(
+      minimumDate
+        ? undefined
+        : sourceMoment.utc().endOf('day').add(1, 'day').toDate(),
+    );
+  };
+
```
Comment (vonovak) [inline example/App.js:127]:
```suggestion
  const toggleMinMaxDate = () => {
    const startOfTodayUTC = sourceMoment.utc().startOf('day').toDate();
    setMinimumDate(
      maximumDate ? undefined : startOfTodayUTC,
    );
    const endOfTomorrowUTC = sourceMoment.utc().endOf('day').add(1, 'day').toDate();
    setMaximumDate(
      minimumDate
        ? undefined
        : endOfTomorrowUTC,
    );
  };

```

File: example/e2e/detoxTest.spec.js [hunk 5]
```diff
@@ -186,6 +186,58 @@ describe('Example', () => {
...
+
+        // Ensure you can select tomorrow (iOS)
+        await userOpensPicker({mode: 'date', display: getPickerDisplay()});
+        await testElement.setDatePickerDate('2021-11-14', 'yyyy-MM-dd');
+      } else {
+        const uiDevice = device.getUiDevice();
+
+        // Ensure you can't select yesterday (Android)
+        const focusTwelethOfNovemberInCalendar = async () => {
+          await uiDevice.pressDPadDown();
```
Comment (vonovak) [inline example/e2e/detoxTest.spec.js:210]: this seems like a simple for loop from 0 to n would already make sense here 🙂 
same thing below

File: example/e2e/detoxTest.spec.js [hunk 6]
```diff
@@ -186,6 +186,58 @@ describe('Example', () => {
       }
       await expect(getTimeText()).toHaveText('09:30');
     });
+
+    it('should let you pick tomorrow but not yesterday when setting min/max', async () => {
+      await element(by.id('DateTimePickerScrollView')).scrollTo('bottom');
```
Comment (vonovak) [inline example/e2e/detoxTest.spec.js:191]: can you please use `elementById` for consistency? thanks

**Review by vonovak — CHANGES_REQUESTED** [review]
hello and thanks so much, I'd like to ask for a few more small changes and then we're good to merge! :)

thanks

**Comment by PaitoAnderson** [comment]
Thank you for reviewing! Let me know if you want anything else :)

**Review by vonovak — APPROVED** [review]

**Comment by vonovak** [comment]
:tada: This issue has been resolved in version 5.0.2 :tada:

If this package helps you, consider [sponsoring us](https://github.com/sponsors/react-native-datetimepicker)! :rocket:

**Pull request merged**